### PR TITLE
i/b/systemd_reload: add systemd-reload interface

### PR DESCRIPTION
**Pull request by alexclewontin**
Alternative proposal to https://github.com/snapcore/snapd/pull/12329, breaking it out into its own super-privileged interface. Name was purely a quick & dirty suggestion, and I'm sure there's a better one.

### CODE REVIEW

**Comment by alexclewontin** [comment]
For the sake of discussion (and continuing a bit off of what @pedronis said in the previous iteration of this PR):

Because 

- historically snapd has a lock on running daemon-reload
- this isn't useful in isolation, but rather in conjunction with a system-files interface to write files to specific locations
- the only real use-case I can see here is running from the install-device hook, to enable the initial reboot from install mode to run mode

would it potentially maybe make more sense to expose a `snapctl daemon-reload` command that could only be called from the install-device hook (similar to `snapctl reboot`)? While the simplest version of this might just do a systemctl daemon-reload, at least it would give snapd the potential to make more complex decisions about whether to allow/combine with another existing daemon-reload/defer until right before reboot/etc

OTOH, that might actually reduce the amount of control we'd have in granting it to people because it wouldn't be "super-privileged". Interested in thoughts

File: interfaces/builtin/systemd_control.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ */
+
+package builtin
+
+const systemdControlSummary = `allows performing certain systemd operations that aren't otherwise managed by snapd`
+
+const systemdControlBaseDeclarationPlugs = `
+  systemd-control:
+    allow-installation: false
```
Comment (pedronis) [inline interfaces/builtin/systemd_control.go:26]: thanks for making it super-privileged

File: interfaces/builtin/systemd_control.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+dbus (send)
+    bus=system
+    path=/org/freedesktop/systemd1
+    interface=org.freedesktop.DBus.Introspectable
+    member=Introspect,
+`
+
+func init() {
+	registerIface(&commonInterface{
+		name:                  "systemd-control",
```
Comment (pedronis) [inline interfaces/builtin/systemd_control.go:61]: I tried to look what else is available on systemd1 and I didn't see anything limited like/like Reload, and given that at this point  afaik we still don't want to have real/full systemd-control interface, we should call this one systemd-reload.

File: interfaces/builtin/systemd_control.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const systemdControlSummary = `allows performing certain systemd operations that aren't otherwise managed by snapd`
```
Comment (pedronis) [inline interfaces/builtin/systemd_control.go:22]: this will need to be scoped, clarified as well to match

**Review by pedronis — COMMENTED** [review]
overall lgtm, comments about the name

File: interfaces/builtin/systemd_control.go (hunk 2 above)
Comment (alexclewontin) [reply interfaces/builtin/systemd_control.go:61]: ACK, will address

File: interfaces/builtin/systemd_control.go (hunk 3 above)
Comment (alexclewontin) [reply interfaces/builtin/systemd_control.go:22]: ACK, will address

**Review by pedronis — APPROVED** [review]
lgtm

**Review by alexmurray — APPROVED** [review]
LGTM but I am a bit worried about @pedronis's comment in the other PR:

> Also snapd itself historically has grown a lock around daemon-reload. Widespread use of daemon-reload open us to probably unlikely but hard to track bugs.

Rather than providing direct access to allow snaps to poke systemd themselves, would it be better to have this functionality mediated by snapd through a `snap systemd-reload` command or similar? Such that a snap which plugs this new `systemd-reload` interface would be allowed to issue such a command and then snapd would do the `systemctl daemon-reload` on its behalf?

**Comment by alexclewontin** [comment]
I'll defer to @pedronis on that question, but happy to take another crack at it if he thinks that's a better way to do it

**Comment by pedronis** [comment]
>  I'll defer to @pedronis on that question, but happy to take another crack at it if he thinks that's a better way to do it

@alexclewontin can you remind me again from where exactly the daemon-reload would be triggered in the know use-case? is this from a hook? which one? the original PR mentions script but scripts don't usually run by themselves, with access

**Comment by alexclewontin** [comment]
@pedronis sorry, should have been more clear on that. 

It needs to be called from the `install-device` hook. The hook also has a system-files plug that looks like

```
plugs:
  systemd-units:
    interface: system-files
    write:
      - /etc/systemd/system
      - /run/mnt/ubuntu-data/system-data/_writable_defaults/etc/systemd/system
```

and writes new services that are `Before={poweroff,reboot,halt}.target`s, respectively. Then it symlinks them into the respective `/run/mnt/ubuntu-data/_writable_defaults/etc/systemd/system/{poweroff,reboot,halt}.target.wants/` directories.

On future boots, this is enough, because now the dependency tree for those targets is loaded correctly initially. The only reason we need to call `daemon-reload` is for that first reboot, from install mode into run mode, as those files of course didn't exist when the dependency tree initially was loaded by systemd.

**Comment by pedronis** [comment]
@alexclewontin 

> It needs to be called from the install-device hook. The hook also has a system-files plug that looks like

the smallest change with minimal long term impact for this would be to have snapd itself do a daemon-reload if there's a install-device or an install-device with system-files from `doRestartSystemToRunMode`,  the handler code for the `restart-system-to-run-mode` task which is executed just after the hook and is the code that triggers a reboot (or whatever else install-device requested via `snapctl reboot`)

**Pull request closed without merging**